### PR TITLE
Add per-node filter queries + remove `_yz_node` field

### PR DESCRIPTION
**Pull request by rzezeski**
Create a custom SearchComponent which allows Yokozuna to perform
per-node filter queries. This removes the need for the `_yz_node` field
and the use of a monolithic filter query. Allowing smaller index size
and less overhead per query.

Preliminary micro benchmark demonstrated 15% improvement in query
throughput. More detailed results to come.
## Benchmark Results

Using my "fruit" micro benchmark I was able to produce a 17% increase
in query throughput from 2665 queries per second to 3124
QPS. Strangely enough there was no reduction in index time or index
disk usage. Perhaps Lucene is really good at optimizing that away
using it's underlying PFOR encoding. I'm honestly not sure but I don't
have the time to dig for more details.
### Indexing

|  | Throughput (WPS) | Median Latency | 99th Latency | Index du |
| --- | --- | --- | --- | --- |
| Monolithic | 1584 | 11.84 ms | 114.10 ms | 41M |
| Per Node | 1593 | 11.57 ms | 114.33 ms | 41M |
### Querying

|  | Throughput (QPS) | Median Latency | 99th Latency |
| --- | --- | --- | --- |
| Monolithic | 2665 | 13.28 ms | 24.03 ms |
| Per Node | 3124 (+ 17%) | 10.62 ms (- 20%) | 22.53 ms (- 6%) |

### CODE REVIEW

File: src/yz_events.erl [hunk 1]
```diff
@@ -82,6 +82,7 @@ handle_info(tick, S) ->
     Current = names(yz_index:get_indexes_from_ring(Ring)),
     {Removed, Added, Same} = yz_misc:delta(Previous, Current),
     ok = sync_indexes(Ring, Removed, Added, Same),
+    remove_non_owned_data(),
```
Comment (coderoshi) [inline src/yz_events.erl]: twice?

Comment (rzezeski) [inline src/yz_events.erl]: Good catch, rebase mistake.

File: java_src/com/basho/yokozuna/handler/component/FQShardTranslator.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        return "https://github.com/basho/yokozuna";
+    }
+
+    private boolean isDistrib(SolrParams params) {
+        // Based on HttpShardHandler because rb.isDistrib is not public.
+        boolean distrib = params.getBool("distrib", false);
+        String shards = params.get(ShardParams.SHARDS);
+        boolean hasShardURL = shards != null;
+
+        return hasShardURL | distrib;
```
Comment (coderoshi) [inline java_src/com/basho/yokozuna/handler/component/FQShardTranslator.java]: A _bitwise or_ (`|`) technically works here, but _boolean or_ (`||`) is the convention when dealing with boolean types. Though despite urban legend, there's no performance difference either way.

Comment (rzezeski) [inline java_src/com/basho/yokozuna/handler/component/FQShardTranslator.java]: Actually `|` is bitwise for primitive integer types, but logical non short circuit OR for boolean. This was copypasta from Solr code, but I agree it should be short circuit OR.

**Comment by coderoshi** [comment]
+1, very cool

security and migration failures are unrelated, so caveat emptor on those results

```
yz_wm_extract_test-bitcask   : pass
yz_stat_test-bitcask         : pass
yz_solr_start_timeout-bitcask: pass
yz_siblings-bitcask          : pass
yz_security-bitcask          : fail
yz_schema_admin-bitcask      : pass
yz_rs_migration_test-bitcask : fail
yz_pb-bitcask                : pass
yz_monitor_solr-bitcask      : pass
yz_mapreduce-bitcask         : pass
yz_languages-bitcask         : pass
yz_index_admin-bitcask       : pass
yz_fallback-bitcask          : pass
yz_errors-bitcask            : pass
yz_dt_test-bitcask           : pass
yokozuna_essential-bitcask   : pass
aae_test-bitcask             : pass
---------------------------------------------
2 Tests Failed
15 Tests Passed
```

**Comment by rzezeski** [comment]
Updated original PR with benchmark details. Test suite has passed twice for me.

**Pull request merged**